Sticky immix

qinsoon opened the pull request:
This PR adds `StickyImmix` as a plan (https://github.com/mmtk/mmtk-core/issues/682). 

Changes:
* Refactoring to current code for sticky plans
  * Introduce more methods to `trait GenerationalPlan` and add `trait GenerationalPlanExt`
  * Refactor `GenObjectBarrierSemantics` and `GenNurseryProcessEdges` so they use the generational plan traits, and work for sticky plans. 
  * Macro-generated post scan will delegate the call to 'parent' if the object is not in any space in the current plan.
  * Extract work scheduling code to `Immix::schedule_immix_full_heap_collection` so all immix-related plans can use it.
  * Add `ImmixSpaceArgs` that changes some Immix behaviors and make it possible to use as a nursery/mature space for sticky immix.
* Add sticky immix
* Refactor `trait Plan` and move some methods to `trait GenerationalPlan`.
* Refactor sanity GC to allow check object sanity with plans and bindings.
* Bug fix: all the plans that use object barrier may have duplicated edges (from memory slice copy barrier)
* Bug fix: clarify non-defrag vs non-moving in Immix (this closes #683)
* Bug fix: `harness_begin` will force a full heap GC (this closes #742)
* Add some logging
  * info log used/total pages before and after GC
  * info log elapsed time for each GC
  * trace log some info about edge tracing, object scan, and a few other places.

caizixian commented on src/plan/immix/global.rs:
```diff
@@ -81,21 +82,25 @@ impl<VM: VMBinding> Plan for Immix<VM> {
-        } else {
-            scheduler.schedule_common_work::<ImmixGCWorkContext<VM, TRACE_KIND_FAST>>(self);
-        }
+        // let in_defrag = self.immix_space.decide_whether_to_defrag(
```
We probably want to remove these commented out sections?

qinsoon replied:
Yes, I will. This is just a draft. If you are curious, you are welcome to look at the changes and leave any comments. But do not spend much time on a detailed review yet.

qinsoon commented:
In terms of performance, I have one run [here (internal url)](https://squirrel.anu.edu.au/plotty/yilin/mmtk/#0|ferret-2023-02-06-Mon-235911&benchmark^buildstring^invocation^iteration&GC^time^time.other^time.stw&|10&iteration^1^2|20&1^invocation|30&1&benchmark&buildstring;jdk-mmtk-master.ix.ms.s.c2.tph.probes_cp.probes_rustmmtk.probes_native.no_biased_locking.no_class_unloading.dacapochopin-0d047f55|41&Histogram%20(with%20CI)^buildstring^benchmark&).
* Basically the sticky immix plan has the same generational behavior as generational immix (they are both faster or slower in a benchmark). 
* Sticky immix is faster than generational immix generally in the measurement, which is strange. It suggests there are some issues with our generational immix implementation -- I do not have time to look into that right now.
* `eclipse` is the only benchmark that sticky immix is slower than generational immix. The issue is that sticky immix cannot properly implement the address check for nursery objects needed the memory slice copy barrier (see the discussion here: https://mmtk.zulipchat.com/#narrow/stream/262673-mmtk-core/topic/sticky-immix-array-copy-barrier/near/327476810)
* `luindex` and `pmd` fail for both generational and sticky immix in both `master` and this PR. I haven't figured out the issue.

I will do a new run with PGO and with new Rust version soon, and post the results here. However, I would like to start the review process right now while waiting for the new performance results.

wks commented on src/plan/immix/global.rs:
```diff
@@ -160,4 +162,37 @@ impl<VM: VMBinding> Immix<VM> {
+        if in_defrag {
+            scheduler.schedule_common_work::<DefragContext>(plan);
+        } else {
+            // The type of plan is `DefragContext::PlanType`, and we need it as `FastContext::PlanType`.
+            // They should be the same plan. But I don't find a way to tell Rust compiler that
+            // those PlanTypes are the same. So just do a unsafe transmute here.
+            let fast_plan = unsafe { std::mem::transmute(plan) };
+            scheduler.schedule_common_work::<FastContext>(fast_plan);
```
We can tell Rust that both `FastContext` and `DefragContext` has the same `PlanType` by adding an additional type parameter.  Here is a patch:

[same-type.patch.gz](https://github.com/mmtk/mmtk-core/files/10729156/same-type.patch.gz)

qinsoon replied:
Thanks. This is very helpful. I will make this change.

wks commented on src/policy/immix/immixspace.rs:
```diff
@@ -511,24 +566,36 @@ impl<VM: VMBinding> ImmixSpace<VM> {
-                ForwardingWord::forward_object::<VM>(object, semantics, copy_context)
+                let new_object =
+                    ForwardingWord::forward_object::<VM>(object, semantics, copy_context);
+                Block::containing::<VM>(new_object).set_state(BlockState::Marked);
```
Both the `if` and the `else` branches set the marked state.  We may merge them and move them after the `if` statement.

qinsoon replied:
I agree. There is an assertion below to check if the block is marked. But the original code does not mark the block if we forward the object. I don't know how it works. We would better check with @wenyuzhao.

wenyuzhao replied:
Blocks are marked when they're allocated from the page-resource or poped out from the reusable blocks queue.  https://github.com/mmtk/mmtk-core/blob/master/src/policy/immix/block.rs#L158-L162 . The store of the mark byte only happens once per block. The `set_state` here is unnecessary and also less efficient. We should remove this.

qinsoon replied:
Thanks @wenyuzhao  for pointing to the code. The assertion failure actually failed for sticky immix. I think the following is the problem.

We have two immix allocators in the worker, one used for defrag (`copy` is `true`), and the other one for normal copy (`copy` is `false`)
https://github.com/mmtk/mmtk-core/blob/106fc71472e0d960e78da977ef2d43d54bbc5113/src/policy/immix/immixspace.rs#L803-L806

For a defrag immix GC, we use the defrag allocator, and the defrag allocator will initiate the block as marked.
https://github.com/mmtk/mmtk-core/blob/106fc71472e0d960e78da977ef2d43d54bbc5113/src/policy/immix/immixspace.rs#L776-L780
https://github.com/mmtk/mmtk-core/blob/106fc71472e0d960e78da977ef2d43d54bbc5113/src/policy/immix/block.rs#L158-L162

For a fast immix GC, we use the normal allocator which does not initiate the blocks as marked, but we use `trace_object_without_moving()` which marks blocks when we trace objects.
https://github.com/mmtk/mmtk-core/blob/106fc71472e0d960e78da977ef2d43d54bbc5113/src/policy/immix/immixspace.rs#L455

So in short, for fast GC, we mark each block at tracing, and for defrag GC, we initiate each block as marked. It is an optimization so we do not need to mark objects if they are copied in the GC.

For sticky immix, this no longer works. A nursery GC in sticky immix is not a defrag GC (so blocks are not initiated as marked), but it may move objects (so it uses `trace_object_with_opportunistic_copying()` which does not mark blocks if an object is copied). So there are some blocks that are not marked if we do not add the line, and we will see the assertion failure.

qinsoon replied:
I will think about what to do for this.

qinsoon replied:
I checked the Java MMTk. And it seems we do not need two different immix copying allocators for most immix plans (only generational immix needs two copying allocators). So I renamed the current `ImmixCopyContext` (with two copying allocators) to `ImmixHybridCopyContext`, and this is used by generational immix. I also added a `ImmixCopyContext` (with one copying allocator), and this is used by immix and sticky immix. In this case, we no longer need the extra block marking in `trace_object`. I will measure performance again and post here.

k-sareen replied:
That's not quite right. StickyImmix uses `nurseryCopy` as well as `copy`. Note that it extends `ImmixCollector` so it already has the `ImmixAllocator copy` field.

qinsoon replied:
I didn't find where `nurseryCopy` is used. Is it used anywhere? It inherits `ImmixCollector`, but does not override `allocCopy`.

k-sareen replied:
Huh -- very odd... You're right. It's initialized but not used. `nurseryCopy.reset()` is called but nothing is ever alloced in it. I'll check git logs to see if it was ever used

k-sareen replied:
It was never used from the initial commit which added Immix (https://github.com/mmtk/jikesrvm/commit/f10ba059ebdb801cc66e2cb6552a6f4a3a0f193d). How odd.

qinsoon replied:
> I checked the Java MMTk. And it seems we do not need two different immix copying allocators for most immix plans (only generational immix needs two copying allocators). So I renamed the current `ImmixCopyContext` (with two copying allocators) to `ImmixHybridCopyContext`, and this is used by generational immix. I also added a `ImmixCopyContext` (with one copying allocator), and this is used by immix and sticky immix. In this case, we no longer need the extra block marking in `trace_object`. I will measure performance again and post here.

The results are [here](https://squirrel.anu.edu.au/plotty/yilin/mmtk/#0|ferret-2023-02-21-Tue-004454&benchmark^build^invocation^iteration&GC^time^time.other^time.stw&|10&iteration^1^2|20&1^invocation|50&GC^1^1^+inf|30&1&benchmark&build;jdk-mmtk-master-ix|41&Histogram%20(with%20CI)^build^benchmark&). I am still running it. For what I have now, the performance looks similar to my previous run (https://github.com/mmtk/mmtk-core/pull/741#issuecomment-1432284832).

wks commented on src/policy/immix/immixspace.rs:
```diff
@@ -646,14 +721,25 @@ pub struct PrepareBlockState<VM: VMBinding> {
+            } else {
+                // If the log bit is not in side metadata, we cannot bulk zero. We can either
+                // clear the bit for dead objects in major GC, or clear the log bit for new
+                // objects. In both cases, we do not need to set log bit at tracing.
```
In English, it is more natural to use "in either case" if the main part is negative, i.e. "we do *not*".

```suggestion
                // objects. In either case, we do not need to set log bit at tracing.
```

See: https://forum.wordreference.com/threads/in-either-case-vs-in-both-case.2243816/

wks commented on src/policy/largeobjectspace.rs:
```diff
@@ -196,9 +196,15 @@ impl<VM: VMBinding> LargeObjectSpace<VM> {
+        trace!(
+            "LOS object {} {} a nursery object",
+            object,
+            if nursery_object { "is" } else { "is not " }
```
```suggestion
            if nursery_object { "is" } else { "is not" }
```

wks commented on src/scheduler/controller.rs:
```diff
@@ -130,6 +126,8 @@ impl<VM: VMBinding> GCController<VM> {
         }
         self.scheduler.deactivate_all();
 
+        let gc_elapsed = std::time::Instant::now() - gc_start;
```
```suggestion
        let gc_elapsed = gc_start.elapsed();
```

wks commented on src/scheduler/gc_work.rs:
```diff
@@ -219,11 +219,16 @@ impl<E: ProcessEdgesWork> GCWork<E::VM> for StopMutators<E> {
 
 #[derive(Default)]
-pub struct EndOfGC;
+pub struct EndOfGC(pub std::time::Duration);
```
I recommend giving the field a name.  It is not very obvious what that duration means when looking at this work packet alone.

wks replied:
> You can't give field names inside of a tuple struct in Rust. Have to make it a proper struct.

Yes. That's what I mean.  Make it a struct with named fields.

qinsoon commented:
I made changes as you suggested @wks, except the module-level comments and `set_state`.

wks reviewed (approved):
LGTM.

qinsoon commented:
> In terms of performance, I have one run [here (internal url)](https://squirrel.anu.edu.au/plotty/yilin/mmtk/#0%7Cferret-2023-02-06-Mon-235911&benchmark%5Ebuildstring%5Einvocation%5Eiteration&GC%5Etime%5Etime.other%5Etime.stw&%7C10&iteration%5E1%5E2%7C20&1%5Einvocation%7C30&1&benchmark&buildstring;jdk-mmtk-master.ix.ms.s.c2.tph.probes_cp.probes_rustmmtk.probes_native.no_biased_locking.no_class_unloading.dacapochopin-0d047f55%7C41&Histogram%20(with%20CI)%5Ebuildstring%5Ebenchmark&).
> 
> * Basically the sticky immix plan has the same generational behavior as generational immix (they are both faster or slower in a benchmark).
> * Sticky immix is faster than generational immix generally in the measurement, which is strange. It suggests there are some issues with our generational immix implementation -- I do not have time to look into that right now.
> * `eclipse` is the only benchmark that sticky immix is slower than generational immix. The issue is that sticky immix cannot properly implement the address check for nursery objects needed the memory slice copy barrier (see the discussion here: https://mmtk.zulipchat.com/#narrow/stream/262673-mmtk-core/topic/sticky-immix-array-copy-barrier/near/327476810)
> * `luindex` and `pmd` fail for both generational and sticky immix in both `master` and this PR. I haven't figured out the issue.
> 
> I will do a new run with PGO and with new Rust version soon, and post the results here. However, I would like to start the review process right now while waiting for the new performance results.

This is the performance run (with Rust 1.66.1 and PGO): [internal link](https://squirrel.anu.edu.au/plotty/yilin/mmtk/#0|ferret-2023-02-14-Tue-050817&benchmark^build^invocation^iteration&GC^time^time.other^time.stw&|10&iteration^1^2|20&1^invocation|50&GC^1^1^+inf|30&1&benchmark&build;jdk-mmtk-master-ix|41&Histogram%20(with%20CI)^build^benchmark&). It is similar to my previous measurement.

Total time

![stickyix-time](https://user-images.githubusercontent.com/1664709/219225674-707267f7-ac68-4b5e-82e9-52fd00faec26.png)

GC time

![stickyix-stw-time](https://user-images.githubusercontent.com/1664709/219226761-a8cff3d1-6e3e-49e0-9731-5da5b0bc1201.png)

wenyuzhao commented on src/mmtk.rs:
```diff
@@ -142,7 +142,10 @@ impl<VM: VMBinding> MMTK<VM> {
-        // FIXME Do a full heap GC if we have generational GC
+        // For generational plans, force a full heap collection.
+        if let Some(gen) = self.plan.generational() {
+            gen.force_full_heap_collection();
```
What about passing another argument to the `handle_user_collection_request` call below to force a full GC?

qinsoon replied:
That sounds good. But we can only add the argument in our internal call, rather than expose the idea of full heap GC to the user.

qinsoon replied:
Done

wenyuzhao commented on src/plan/global.rs:
```diff
@@ -678,7 +693,9 @@ impl<VM: VMBinding> BasePlan<VM> {
         if emergency_collection {
-            plan.force_full_heap_collection();
+            if let Some(gen) = plan.generational() {
+                gen.force_full_heap_collection();
```
it's possible to call `plan.force_full_heap_collection();` all the time. For non-generational full-heap collectors this is just a no-op.

qinsoon replied:
That is what the current code does, and `force_full_heap_collection()` is a method of `trait Plan`. I moved the method to `trait GenerationalPlan`. I think it is cleaner, as not all plans are generational, and the idea of full heap does not apply to them.

wenyuzhao commented on src/plan/generational/immix/global.rs:
```diff
@@ -225,11 +236,13 @@ impl<VM: VMBinding> GenImmix<VM> {
+            plan_args.get_space_args("immix_mature", true, VMRequest::discontiguous()),
+            ImmixSpaceArgs {
+                reset_log_bit_in_major_gc: false,
+                unlog_object_when_traced: false,
```
shouldn't this be true?

qinsoon replied:
There is a slight difference.

In generational immix, we only need to unlog objects in GC when we copy an object to the mature/immix space (no matter from nursery or from the mature space). When we copy from nursery to mature, we do not invoke `ImmixSpace::trace_object`. So setting `unlog_object_when_traced` to `true` will not affect the log bit for those promoted object (as `ImmixSpace::trace_object` is not called on them). We use `CopySemantics::PromoteToMature` or `CopySemantics::Mature` to set the log bit in the `post_copy`.

In sticky immix, any object we trace in the immix space (no matter copied or not) is considered promoted, and we need to unlog it. For example, when we trace a nursery object, even if we do not copy it, it is 'promoted' to a mature object, and we need to unlog it. In this case, using `post_copy` won't capture all the cases here. So I added this option, and for sticky immix, we can just unlog every object when we trace it.

qinsoon replied:
I can add some comments to make it clear why we do not need it to be `true`.

qinsoon replied:
I added some comments to explain this.

wenyuzhao commented on src/plan/generational/global.rs:
```diff
@@ -298,21 +300,53 @@ impl<VM: VMBinding> CommonGenPlan<VM> {
+    fn is_current_gc_nursery(&self) -> bool;
+
+    /// Is the object in the nursery?
+    fn is_object_in_nursery(&self, object: ObjectReference) -> bool;
```
This function only works based on the assumption that we only have two generations in the heap (i.e. young-gen and old-gen). If the number of generations is larger than 2, shall we do something different to handle this?

As an example, G1 has three generations: eden, survived, and old.

qinsoon replied:
If we need to support multiple generations, we will have to do some refactoring to allow it. For example, `gen_plan.is_object_in_nursery(object, Generationa::Eden)`. Alternatively, we may use a new trait, like `MultiGenerationalPlan`. I don't know what would be a better way.

The changes in this PR is to allow us to reuse the generational code for sticky plans. I think if we want to support multiple generations, we would also want to reuse the current generational code as much as we can.

wenyuzhao commented on src/policy/immix/immixspace.rs:
```diff
@@ -511,24 +566,36 @@ impl<VM: VMBinding> ImmixSpace<VM> {
             );
             queue.enqueue(new_object);
             debug_assert!(new_object.is_live());
+            self.unlog_object_if_needed(new_object);
```
A better approach is to bulk zero the unlog bits when a clean block is allocated, or when reusing a reusable hole, instead of unlog each object here. Unlog every individual object may not be faster than bulk zeroing.

qinsoon replied:
Can you elaborate on this a bit more? I am confused. Bulk zeroing is setting bits to 0, but here we set the bit to 1 (making the object a mature object). At the end of a nursery GC, all the mature objects should have bits as 1.

Also bear in mind that in sticky immix, an object becomes mature as long as we trace it in a nursery GC (we do not have to copy it to make it a mature object). So usually we unlog an object when we copy it, but in sticky immix, we need to unlog an object when we trace it.

wenyuzhao replied:
>  Bulk zeroing is setting bits to 0, but here we set the bit to 1 (making the object a mature object)

yes I mean bulk-memset the unlog table to 1

> So usually we unlog an object when we copy it, but in sticky immix, we need to unlog an object when we trace it.

For object left in place without moving, yes you have to unlog them without bulk-memset.

But bulk-memset can be applied to to-space. For this case, the object is already copied. The to-space memory should be bulk initialized to 1 in allocation slow-path.

wenyuzhao replied:
Say if you're copying 1k promoted objects to a 32 KB clean block, bulk initializing the unlog table once is definitely faster then setting the unlog table 1k times for each copied object.

qinsoon replied:
The thing is that we may not copy an object at all in sticky immix. The object is allocated with `log bit === 0`, but when we trace it, we need to set its log bit to 1. However, I can try this: for objects that are copied, bulk set their log bit, for objects that are not copied, set the log bit per object.

wenyuzhao commented on src/policy/immix/immixspace.rs:
```diff
@@ -646,14 +721,25 @@ pub struct PrepareBlockState<VM: VMBinding> {
+        if self.space.space_args.reset_log_bit_in_major_gc {
+            if let MetadataSpec::OnSide(side) = *VM::VMObjectModel::GLOBAL_LOG_BIT_SPEC {
+                // We zero all the log bits in major GC, and for every object we trace, we will mark the log bit again.
+                side.bzero_metadata(self.chunk.start(), Chunk::BYTES);
```
When do you need to bulk zero all the unlog bits in the heap? This sounds expensive.

qinsoon replied:
`PrepareBlockState` is only scheduled for full heap GCs. And this bulk zeroing is only for sticky immix, it does not affect other immix variants.

wenyuzhao replied:
Sorry I mean for sticky immix, why all the unlog table need to be reset before a full-heap GC?

qinsoon replied:
We need a way to set the log bit back to `0` for dead objects after a full heap GC. We can either clear the log bit when we allocate new objects in mutator phase, or just bulk zero the log bit and set the bit for the live objects. As I already need to set the bit for live objects (see our discussion in the other comment), I think it is natural to just bulk zero the log bit.

qinsoon commented:
@wenyuzhao I have addressed your comments. I am running performance measurement for the changes in https://github.com/mmtk/mmtk-core/pull/741#discussion_r1112381496. I am not doing other work for this PR right now. If there is any other requested change, just let me know. Otherwise this PR will be merged if the new performance results are similar to the previous one (https://github.com/mmtk/mmtk-core/pull/741#issuecomment-1432284832) after I get them.

wenyuzhao commented:
> @wenyuzhao I have addressed your comments. I am running performance measurement for the changes in [#741 (comment)](https://github.com/mmtk/mmtk-core/pull/741#discussion_r1112381496). I am not doing other work for this PR right now. If there is any other requested change, just let me know. Otherwise this PR will be merged if the new performance results are similar to the previous one ([#741 (comment)](https://github.com/mmtk/mmtk-core/pull/741#issuecomment-1432284832)) after I get them.

All other parts looks good to me. I'm only concerning the performance of the unlog table related code. Mainly two parts: (1) setting unlog bits when trace objects (2) bulking zeroing all the unlog table before a full GC. These do not affect correctness, but are performance critical.

qinsoon commented:
> > @wenyuzhao I have addressed your comments. I am running performance measurement for the changes in [#741 (comment)](https://github.com/mmtk/mmtk-core/pull/741#discussion_r1112381496). I am not doing other work for this PR right now. If there is any other requested change, just let me know. Otherwise this PR will be merged if the new performance results are similar to the previous one ([#741 (comment)](https://github.com/mmtk/mmtk-core/pull/741#issuecomment-1432284832)) after I get them.
> 
> All other parts looks good to me. I'm only concerning the performance of the unlog table related code. Mainly two parts: (1) setting unlog bits when trace objects (2) bulking zeroing all the unlog table before a full GC. These do not affect correctness, but are performance critical.

I can try this implementation and see how it performs:
* bulk set log bit to 1 in copying
* set log bit to 1 per object if an object is not copied
* bulk set log bit to 0 in allocation

Is this what you suggest?

qinsoon commented:
> > > @wenyuzhao I have addressed your comments. I am running performance measurement for the changes in [#741 (comment)](https://github.com/mmtk/mmtk-core/pull/741#discussion_r1112381496). I am not doing other work for this PR right now. If there is any other requested change, just let me know. Otherwise this PR will be merged if the new performance results are similar to the previous one ([#741 (comment)](https://github.com/mmtk/mmtk-core/pull/741#issuecomment-1432284832)) after I get them.
> > 
> > 
> > All other parts looks good to me. I'm only concerning the performance of the unlog table related code. Mainly two parts: (1) setting unlog bits when trace objects (2) bulking zeroing all the unlog table before a full GC. These do not affect correctness, but are performance critical.
> 
> I can try this implementation and see how it performs:
> 
> * bulk set log bit to 1 in copying
> * set log bit to 1 per object if an object is not copied
> * bulk set log bit to 0 in allocation
> 
> Is this what you suggest?

We discussed this in today's meeting:
* Bulk setting log bits to 1 may not be necessary for object remembering barriers. But it could be an important optimisation for field remembering barriers.
* For object remembering barriers, we could simply set the byte to `0xff` rather than masking and setting bits. It will set adjacent bits to 1, but it is still correctness and it improves performance.
* With the discussion above, we will need to unlog every object we trace in a nursery GC, and we need to unlog every 'nursery' object in a mature GC as well. To make it simple, we can just zero all the log bits in mature GC, and rely on the same code path in object tracing to unlog every object in a mature GC. An alternative approach is to bulk zero log bit in allocation (lazily) or in releasing (eagerly) (but we still need to unlog nursery objects in tracing). This could be more efficient.

What I will do:
* Set byte to unlog objects
* Try bulk zeroing log bit for dead objects after major GC (instead of bulk zeroing log bits for all blocks in major GC).

qinsoon commented:
This is the PR for setting byte for log bit: https://github.com/mmtk/mmtk-core/pull/764. I measured with GenImmix, and it seems no measurable difference in GC time.

I don't think it works with sticky immix, as in sticky immix, nursery objects can be adjacent to mature objects. If we simply set byte for log bit in a GC, the new allocated nursery objects may have its log bit as `1`.

wenyuzhao commented:
> Also I don't think it works with sticky immix, as in sticky immix, nursery objects can be adjacent to mature objects. If we simply set byte for log bit in a GC, the new allocated nursery objects may have its log bit as 1.

I don't think this is true. Nursery and mature objects cannot sit in the same line. A line is 256 B, which maps to 32 bits in the unlog table (if the min object size is 8 bytes). Relax-storing a byte (or even an `i32`) should be safe.

qinsoon commented:
> > Also I don't think it works with sticky immix, as in sticky immix, nursery objects can be adjacent to mature objects. If we simply set byte for log bit in a GC, the new allocated nursery objects may have its log bit as 1.
> 
> I don't think this is true. Nursery and mature objects cannot sit in the same line. A line is 256 B, which maps to 32 bits in the unlog table (if the min object size is 8 bytes). Relax-storing a byte (or even an `i32`) should be safe.

I can try and see if it works/improves performance for sticky immix. From my previous runs, this change does not improve performance for gen immix.

qinsoon commented:
> What I will do:
> 
> * Set byte to unlog objects

There are some links for the results. It is still running. 
* Sticky immix: [internal link](https://squirrel.anu.edu.au/plotty/yilin/mmtk/#0|fisher-2023-02-24-Fri-022521&benchmark^buildstring^invocation^iteration&time^time.other^time.stw&|10&iteration^1^2|10&buildstring^2^jdk-mmtk-log-bit-set-byte.genix.ms.s.c2.tph.probes_cp.probes_rustmmtk.probes_native.no_biased_locking.no_class_unloading.dacapochopin-0d047f55|10&buildstring^2^jdk-mmtk-log-bit-set-bit.genix.ms.s.c2.tph.probes_cp.probes_rustmmtk.probes_native.no_biased_locking.no_class_unloading.dacapochopin-0d047f55|20&1^invocation|30&1&benchmark&buildstring;jdk-mmtk-log-bit-set-bit.stickyix.ms.s.c2.tph.probes_cp.probes_rustmmtk.probes_native.no_biased_locking.no_class_unloading.dacapochopin-0d047f55|41&Histogram%20(with%20CI)^buildstring^benchmark&)
* Gen immix: [internal link](https://squirrel.anu.edu.au/plotty/yilin/mmtk/#0|fisher-2023-02-24-Fri-022521&benchmark^buildstring^invocation^iteration&time^time.other^time.stw&|10&iteration^1^2|10&buildstring^2^jdk-mmtk-log-bit-set-byte.stickyix.ms.s.c2.tph.probes_cp.probes_rustmmtk.probes_native.no_biased_locking.no_class_unloading.dacapochopin-0d047f55|10&buildstring^2^jdk-mmtk-log-bit-set-bit.stickyix.ms.s.c2.tph.probes_cp.probes_rustmmtk.probes_native.no_biased_locking.no_class_unloading.dacapochopin-0d047f55|20&1^invocation|30&1&benchmark&buildstring;jdk-mmtk-log-bit-set-bit.genix.ms.s.c2.tph.probes_cp.probes_rustmmtk.probes_native.no_biased_locking.no_class_unloading.dacapochopin-0d047f55|41&Histogram%20(with%20CI)^buildstring^benchmark&)

> * Try bulk zeroing log bit for dead objects after major GC (instead of bulk zeroing log bits for all blocks in major GC).

This is the link for the results (still running): [internal link](https://squirrel.anu.edu.au/plotty/yilin/mmtk/#0|ferret-2023-02-24-Fri-022105&benchmark^build^invocation^iteration&time^time.other^time.stw^work.PlanProcessEdges.time.total^work.PrepareBlockState.time.total^work.SweepChunk.time.total&|10&iteration^1^2|20&1^invocation)

I will update this post once I get results. If the results are positive, I will adopt those changes. Otherwise, I do not plan to spend more time on this PR. I don't think there is any major performance issue for this PR. If we do want to optimize further, we can do it in separate PRs.

qinsoon commented:
binding-refs
JULIA_BINDING_REPO=qinsoon/mmtk-julia
JULIA_BINDING_REF=add-immix-no-nursery-copy
